### PR TITLE
Support getting latest version of dataset

### PR DESCRIPTION
**Pull request by SophieGuo410**

### CODE REVIEW

**Comment by snalli** [comment]
will get back tmrw

File: ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java [hunk 1]
```diff
@@ -137,8 +138,12 @@ public AccountDao(MySqlDataAccessor dataAccessor) {
         String.format("select %s, %s, %s, %s, %s, %s from %s where %s = ? and %s = ? and %s = ?", DATASET_NAME,
             VERSION_SCHEMA, LAST_MODIFIED_TIME, RETENTION_COUNT, USER_TAGS, DELETE_TS, DATASET_TABLE, ACCOUNT_ID,
             CONTAINER_ID, DATASET_NAME);
-    insertDatasetVersionSql = String.format("insert into %s (%s, %s, %s, %s, %s, %s) values (?, ?, ?, ?, now(3), ?)",
-        DATASET_VERSION_TABLE, ACCOUNT_ID, CONTAINER_ID, DATASET_NAME, VERSION, LAST_MODIFIED_TIME, DELETE_TS);
+    listVersionSql =
+        String.format("select %1$s from %2$s " + "where (%3$s, %4$s, %5$s) = (?, ?, ?) ORDER BY %1$s DESC LIMIT 1",
```
Comment (snalli) [inline ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java:142]: why is there a concatenation here ?

File: ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java [hunk 2]
```diff
@@ -448,6 +453,53 @@ public synchronized Dataset getDataset(int accountId, int containerId, String ac
...
+   */
+  private long executeListVersionStatement(PreparedStatement statement, int accountId, int containerId,
+      String datasetName) throws SQLException {
+    ResultSet resultSet = null;
+    try {
+      statement.setInt(1, accountId);
+      statement.setInt(2, containerId);
+      statement.setString(3, datasetName);
+      resultSet = statement.executeQuery();
+      resultSet.next();
```
Comment (snalli) [inline ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java:496]: what if the resultSet is empty for some reason ?

File: ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java [hunk 3]
```diff
@@ -448,6 +453,53 @@ public synchronized Dataset getDataset(int accountId, int containerId, String ac
...
+   * @throws SQLException
+   */
+  private long executeListVersionStatement(PreparedStatement statement, int accountId, int containerId,
+      String datasetName) throws SQLException {
+    ResultSet resultSet = null;
+    try {
+      statement.setInt(1, accountId);
+      statement.setInt(2, containerId);
+      statement.setString(3, datasetName);
+      resultSet = statement.executeQuery();
```
Comment (snalli) [inline ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java:495]: is there a way to specify timeouts for these queries ?

File: ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java (hunk 2 above)
Comment (snalli) [reply ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java:496]:
```
        if (!resultSet.next()) {
          throw Exception;
        }
```

File: ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java (hunk 1 above)
Comment (SophieGuo410) [reply ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java]: Updated.

File: ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java (hunk 2 above)
Comment (SophieGuo410) [reply ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java]: Thanks, fixed.

File: ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java (hunk 3 above)
Comment (SophieGuo410) [reply ambry-account/src/main/java/com/github/ambry/account/mysql/AccountDao.java:495]: I think we don't need to specify right now,  we have the mysql connection timeout, and if we hit the timeout, we can work with mysql team to tune it.

**Review by Arun-LinkedIn — APPROVED** [review]

**Comment by snalli** [comment]
lgtm, you can merge when u feel its ready

**Review by snalli — APPROVED** [review]

**Review by Arun-LinkedIn — APPROVED** [review]

**Pull request merged**